### PR TITLE
LiveEffect: Add foreground service

### LINKED ISSUE #2062: Glitch in background, exclusive mode
Hello.

I'm working on a plugin that should be open when my application runs in background.

I use a fullduplex stream, as i filter the input from the default microphone. The idea is to keep the app alive while recording the surrounding sound.

For now, all is ok when I run my plugin from the app. But when going in background the stream becomes glitchy.

I'm not sure also to really be able to be able to start a stream with exclusive mode. Do I need to do something special?

Thank you!

### PR DESCRIPTION
**Pull request by robertwu1**
Fixes #2062

### CODE REVIEW

File: samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/DuplexStreamForegroundService.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            case ACTION_START:
+                Log.i(TAG, "Receive ACTION_START" + intent.getExtras());
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                    startForeground(1, buildNotification(),
+                            ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PLAYBACK
+                                    | ServiceInfo.FOREGROUND_SERVICE_TYPE_MICROPHONE);
+                }
+                break;
+            case ACTION_STOP:
+                Log.i(TAG, "Receive STOP_RECORD" + intent.getExtras());
```
Comment (philburk) [inline samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/DuplexStreamForegroundService.java]: Should STOP_RECORD be ACTION_STOP?

File: samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/DuplexStreamForegroundService.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        switch (intent.getAction()) {
+            case ACTION_START:
+                Log.i(TAG, "Receive ACTION_START" + intent.getExtras());
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                    startForeground(1, buildNotification(),
+                            ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PLAYBACK
+                                    | ServiceInfo.FOREGROUND_SERVICE_TYPE_MICROPHONE);
+                }
+                break;
+            case ACTION_STOP:
```
Comment (philburk) [inline samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/DuplexStreamForegroundService.java:83]: It's funny that the ACTION_STOP is handled by onStartCommand().

File: samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/MainActivity.java [hunk 3]
```diff
@@ -163,6 +172,12 @@ protected void onPause() {
     @Override
     protected void onDestroy() {
         onStopTest();
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+            Intent serviceIntent = new Intent(ACTION_START, null, this,
```
Comment (philburk) [inline samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/MainActivity.java]: Should this be ACTION_STOP?

**Review by philburk — COMMENTED** [review]
Does the service stop while it is in the background?
Can the audio keep running?

File: samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/MainActivity.java (hunk 3 above)
Comment (robertwu1) [reply samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/MainActivity.java]: Fixed in a newer patch

File: samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/DuplexStreamForegroundService.java (hunk 2 above)
Comment (robertwu1) [reply samples/LiveEffect/src/main/java/com/google/oboe/samples/liveEffect/DuplexStreamForegroundService.java:83]: Yea, it's odd but onStartCommand() handles all commands like how a command queue would work.

**Review by philburk — APPROVED** [review]

**Pull request merged**